File: cpp/15. [M] 3Sum.cpp
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;;

class Solution {
public:
    vector<vector<int>> ThreeSumV1(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int size = nums.size();
        int i = 0;
        vector<vector<int>> res = {};
        // for (auto itr = nums.begin(); itr != nums.end(); itr++) {
        //     cout << *itr << ' ';
        // }
        // cout << endl;

        while (i + 2 < size) {
            int l = i + 1;
            int r = size - 1;
            int sum = nums[i] + nums[l] + nums[r];

            // cout << i << ' ' << l << ' ' << r << endl;
            while (l < r) {
                sum = nums.at(i) + nums.at(l) + nums.at(r);
                if (sum == 0) {
                    // cout << "success" << endl;
                    res.push_back({nums.at(i), nums.at(l), nums.at(r)});
                    int old = nums.at(l);
                    while (l < r && nums.at(l) == old)
                        l++;
                    old = nums.at(r);
                    while (l < r && nums.at(r) == old)
                        r--;
                } else if (sum < 0) {
                    int old = nums.at(l);
                    while (l < r && nums.at(l) == old)
                        l++;
                } else {
                    int old = nums.at(r);
                    while (l < r && nums.at(r) == old)
                        r--;
                }
            }
            // cout << i << ' ' << l << ' ' << r << endl;
            int old = nums.at(i);
            while (i < size && nums.at(i) == old)
                i++;
        }
        return res;
    }

    vector<vector<int>> ThreeSumV2(const vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        int nums_size = nums.size();
        for (int i = 0; i < nums_size; i++) {
            if (i > 0 && nums.at(i) == nums.at(i-1))
                continue;
            int first = nums.at(i);
            int second_i = i + 1;
            int third_i = nums_size - 1;
            while (second_i < third_i) {
                if (second_i > i + 1 && nums.at(second_i) == nums.at(second_i-1)) {
                    second_i++;
                    continue;
                }
                if (third_i < nums_size - 1 && nums.at(third_i) == nums.at(third_i+1)) {
                    third_i--;
                    continue;
                }
                int curr_sum = first + nums.at(second_i) + nums.at(third_i);
                if (curr_sum == 0) {
                    res.push_back({first, nums.at(second_i), nums.at(third_i)});
                    second_i++, third_i--;
                } else if (curr_sum < 0) {
                    second_i++;
                } else {
                    third_i--;
                }
            }
        }
        return res;
    }
};

int main() {
    Solution sol = Solution();
    vector<int> v = {1,2,3,4,1,1,-1,-1,-2};
    v = {0,0,0};
}

File: python/15. [M] 3Sum.py
# https://leetcode-cn.com/problems/3sum/

# Given an integer array nums, return all the triplets
# [nums[i], nums[j], nums[k]] such that i != j, i != k,
# and j != k, and nums[i] + nums[j] + nums[k] == 0.
#
# Notice that the solution set must not contain duplicate triplets.
#
# 0 <= nums.length
#
# 来源：力扣（LeetCode）
# 链接：https://leetcode-cn.com/problems/3sum
# 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。

from typing import List


class Solution:

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        length = len(nums)
        for left_i, left in enumerate(nums):
            if left_i > 0 and nums[left_i - 1] == left:
                continue
            mid_i, right_i = left_i + 1, length - 1
            while mid_i < right_i:
                mid, right = nums[mid_i], nums[right_i]
                if left + mid + right <= 0:
                    if left + mid + right == 0:
                        res.append([left, mid, right])
                    mid_i += 1
                    while mid_i < length and nums[mid_i] == nums[mid_i - 1]:
                        mid_i += 1
                else:
                    right_i -= 1
                    while right_i >= 0 and nums[right_i] == nums[right_i + 1]:
                        right_i -= 1
        return res

    def threeSum_v0(self, nums: list[int]) -> list[list[int]]:
        res = []
        nums.sort()
        ptr = 0
        n = len(nums)

        # set the first ptr in each loop
        while ptr < n and nums[ptr] <= 0:
            l, r = ptr + 1, n - 1
            while l < r:
                # if result found
                if nums[l] + nums[r] == -nums[ptr]:
                    res.append([nums[ptr], nums[l], nums[r]])
                    while l + 1 < r and nums[l] == nums[l + 1]:
                        l += 1
                    l += 1
                    while r - 1 > l and nums[r] == nums[r - 1]:
                        r -= 1
                    r -= 1

                # if twoSum too small
                elif nums[l] + nums[r] < -nums[ptr]:
                    while l + 1 < r and nums[l] == nums[l + 1]:
                        l += 1
                    l += 1

                # if twoSUm too large
                elif nums[l] + nums[r] > -nums[ptr]:
                    while r - 1 > l and nums[r] == nums[r - 1]:
                        r -= 1
                    r -= 1

            # skip through duplicates
            while ptr + 1 < n and nums[ptr] == nums[ptr + 1]:
                ptr += 1
            ptr += 1

        return res

    def threeSum_try2(self, nums: list[int]) -> list[list[int]]:

        def skip_duplicates(start, direction, leng) -> int:
            curr = nums[start]
            while leng > start > 0 and nums[start] == curr:
                start = start + direction
            return start

        LEFT, RIGHT = -1, 1
        leng = len(nums)
        if leng < 3: return []
        res = []
        nums.sort()
        first = 0

        while first < leng - 2:
            second, third = first + 1, leng - 1
            while second < third:
                summ = nums[first] + nums[second] + nums[third]
                if summ == 0:
                    res.append([nums[first], nums[second], nums[third]])
                    second = skip_duplicates(second, RIGHT, leng)
                    third = skip_duplicates(third, LEFT, leng)
                elif summ < 0:
                    second += 1
                else:
                    third -= 1
            curr = nums[first]
            # move the first pointer skipping duplicate values
            while first < leng and nums[first] == curr:
                first += 1
        return res


if __name__ == '__main__':
    sol = Solution()
    print(sol.threeSum([-1, 0, 1, 2, -1, -4]))
    # print(sol.threeSum_try2([-1,-1,1,2,-1,-3]))
    # assert(sol.threeSum_try2([-1,-1,1,2,-1,-1]) == [[-1,-1,2]])
    # assert(sol.threeSum_try2([1]) == [])
    # assert(sol.threeSum_try2([1,2]) == [])
    # assert(sol.threeSum_try2([]) == [])
